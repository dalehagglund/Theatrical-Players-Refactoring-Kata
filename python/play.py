class Play:
    def __init__(self, id: str, name: str, type: str):
        if type not in ("tragedy", "comedy"):
            raise ValueError(f'unknown type: {type}')
        self._id: str = id
        self._name: str = name
        self._type: str = type
        
    def id(self) -> str: return self._id
    def name(self) -> str: return self._name

    def performance_cost(self, audience: int) -> int:
        if self._type == "tragedy":
            return 40000 + 1000 * max(0, audience - 30)
        elif self._type == "comedy":
            return (
                30000 
                + (300 * audience)
                + (10000 if  audience > 20 else 0)
                + 500 * max(0, audience - 20)
            )
        else:
            raise ValueError(f'unknown type: {self.type}')
            
    def volume_credits(self, audience: int) -> int:
        credits = 0
        credits += max(audience - 30, 0)
        if self._type == "comedy":
            credits += audience // 5
        return credits

