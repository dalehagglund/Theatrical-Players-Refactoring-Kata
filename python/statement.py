import math
from play import Play
from typing import Protocol

class InvoiceFormatter(Protocol):
    pass
    
class TextFormatter(InvoiceFormatter):
    pass
    
def statement(format: InvoiceFormatter, invoice, plays):
    total_amount = 0
    volume_credits = 0
    result = f'Statement for {invoice["customer"]}\n'
    
    plays = {
        id: Play(id, info["name"], info["type"])        
        for id, info in plays.items()
    }

    def format_as_dollars(amount):
        return f"${amount:0,.2f}"

    for perf in invoice['performances']:
        play = plays[perf['playID']]
        
        this_amount = play.performance_cost(perf['audience'])
        volume_credits += play.volume_credits(perf['audience'])

        result += f' {play.name()}: {format_as_dollars(this_amount/100)} ({perf["audience"]} seats)\n'
        total_amount += this_amount

    result += f'Amount owed is {format_as_dollars(total_amount/100)}\n'
    result += f'You earned {volume_credits} credits\n'
    return result


