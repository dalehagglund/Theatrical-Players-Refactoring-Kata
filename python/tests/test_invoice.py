from statement import Invoice, InvoiceLine, format_as_html
from play import Play
from approvaltests import verify

hamlet = Play("ham", "Hamlet: Prince of Denmark", "tragedy")

def test_empty_invoice():
    inv = Invoice("customer")
    assert inv.customer() == "customer"
    assert len(list(inv.line_items())) == 0
    assert inv.total_cost() == 0
    assert inv.total_credits() == 0
    
def test_single_performance_has_one_line_item():
    inv = Invoice("customer")
    inv.add_performance(hamlet, 0)
    assert len(list(inv.line_items())) == 1

def test_single_performance_has_correct_totals():
    inv = Invoice("customer")
    inv.add_performance(hamlet, 0)
    assert inv.total_cost() == sum(item.cost for item in inv.line_items())
    assert inv.total_credits() == sum(item.credits for item in inv.line_items())
    
def test_html_formatter():
    inv = Invoice("Looney Tunes Playhouse")
    inv.add_performance(hamlet, 0)
    output = format_as_html(inv)
    verify(output)