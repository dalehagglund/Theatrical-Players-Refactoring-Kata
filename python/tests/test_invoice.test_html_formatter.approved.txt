<html>
<body>
<h1>
Statement for Looney Tunes Playhouse
</h1>
<table>
<tr>
<th>
Play</th>
<th>
Cost</th>
<th>
Audience</th>
</tr>
<tr>
<td>
Hamlet: Prince of Denmark</td>
<td>
40000</td>
<td>
0</td>
</tr>
</table>
<p>
Amount Owed is: 40000
<br/>You've earned 0 credits
</p>
</body>
</html>
